### DISCUSSION TITLE
conda-lock

### DISCUSSION THREAD

**Discussion by MattF-NSIDC** [nsidc/tech-radar#41] [Tools]
`conda env export > environment-lock.yml` is an improvement over no lockfile, but has some disadvantages:

* Platform-dependent: The lockfile will be for whatever platform you ran the command on. Not friendly to MacOS developers who may break CI by locking the env on their box.
* Computer-dependent: The final line of a lock file is `prefix: /prefix/path/for/my/local/machine`. This is noisy and not useful, and we often discard the line with `| grep -v "^prefix:"`.
* No validation: Can't be sure your environment is exactly the same as last time. If someone replaced a package with an altered version (with same version/build numbers), you wouldn't know about it.
* You have to have a live environment to create a lockfile. If you removed packages from `environment.yml` but didn't recreate that environment, your lockfile will have those undesired packages in it.

`conda-lock` solves these problems:

* Creates multi-platform lock-files by default. All platforms are encoded in the same lockfile.
* Has no computer-dependent content
* Hashes all packages in the environment and records those hashes in the lockfile for validation when re-creating the environment
* Can create a lockfile from a conda environment YAML file _even if there is no installed environment_. If you remove a package from your `environment.yml`, you can immediately lock it without having to also remove that package from your live environment. 
* Previously, conda-lock did not support pip packages, but now it does (v1.0.0)
* Previously, conda-lock created separate lock files for each platform, but now they are unified (v1.0.0)

Example migration PR: https://github.com/nsidc/qgreenland/pull/545/files#diff-63113c19c5d310b8e350b302f279e2297ba6faa9ac9c99c1c82e83e508447865

**Comment by rmarow** [comment]
Also using this on the USAON project - https://github.com/nsidc/usaon-vta-survey
Really enjoying the experience - was very easy to integrate with Docker container as well.
